### PR TITLE
Allow to pass css selectors for `skip_area` and `crop`

### PR DESCRIPTION
**Pull request by pftg**
## (New Feature) Allows to provide css selectors as coordinates for `skip_area` and `crop` arguments

`screenshot 'cropped_form', crop: 'form'` - to crop first visible form `<form>`

`screenshot 'page_ignore_footer', skip_area: '#footer'` - to skip an element with `footer` id 

## Refactoring

Use `Region` class instead of pure coordinates in the comparator and drivers

### CODE REVIEW

**Comment by pftg** [comment]
Hey, @donv please check this PR. I'll add some documentation soon.

**Review by pftg — COMMENTED** [review]
Adds readme

**Comment by pftg** [comment]
@UweKubosch please check

**Comment by donv** [comment]
Wow, big change!  It will make my next merge hard :|  

I like the change overall.

Chunky PNG #rect does **not** use width+height, so the change for areas is not a given, and I don't like adding breaking changes with just a personal preference as a reason.  However, I don't feel strongly about it, and it seems you do, so it is fine by me.

Did you change the return value for differences from `nil` to `false` when the old image is missing or the dimensions have changed?  I feel it is useful to distinguish the case from a regular difference.

I am prepared to merge this if you are happy with the above comments.

**Comment by pftg** [comment]
@donv 

1. We will try to drop the breaking change, for now, until guys will migrate to CSS selectors;
2. `nil` for `different?` I found that there is more work to clean up. I'd better drop `nil` for now, will check if there is a usage of the `nil`. Later I want to add more advanced comparison states (like validation errors in rails), in order to prevent double computations in `quick_equal?` and `different?`

File: lib/capybara/screenshot/diff/drivers/chunky_png_driver.rb [hunk 1]
```diff
@@ -183,10 +184,11 @@ def dimension_changed?(old_image, new_image)
             true
           end
 
-          def draw_rectangles(images, (left, top, right, bottom), (r, g, b))
+          def draw_rectangles(images, (left, top, width, height), (r, g, b))
             images.map do |image|
               new_img = image.dup
-              new_img.rect(left - 1, top - 1, right + 1, bottom + 1, ChunkyPNG::Color.rgb(r, g, b))
+              new_img.rect(left - 1, top - 1, left + width + 1, top + height + 1, ChunkyPNG::Color.rgb(r, g, b))
```
Comment (pftg) [inline lib/capybara/screenshot/diff/drivers/chunky_png_driver.rb:190]: @donv yep, sorry for misguiding, ChunkyPNG use `[right, bottom]` style. But JavaScript return rectangle as `[width, height]` style

**Comment by UweKubosch** [comment]
Are you ready to merge now?

**Comment by pftg** [comment]
@donv nope, I'll add updates today

File: lib/capybara/screenshot/diff/drivers/chunky_png_driver.rb (hunk 1 above)
Comment (pftg) [reply lib/capybara/screenshot/diff/drivers/chunky_png_driver.rb]: but for `crop` it uses `[width, height]`, hm strange

**Comment by pftg** [comment]
@donv updated, dropped breaking change. Added more refactoring to use the `Region` class instead of `array` for passing coordinates.

**Comment by pftg** [comment]
@donv @UweKubosch when could you check the last updates?

**Comment by donv** [comment]
Checking them now 😄

File: README.md [hunk 2]
```diff
@@ -452,25 +447,25 @@ Capybara::Screenshot::Diff.area_size_limit = 42
...
 
 Sometimes you have expected change that you want to ignore.
-You can use the `skip_area` option to the `screenshot` method to ignore an area:
+You can use the `skip_area` option with `[left, top, right, bottom]` or css selector like `'body #skiped_element'' to the `screenshot` method to ignore an area:
 
 ```ruby
 test 'unstable area' do
   visit '/'
-  screenshot 'index', skip_area: [17, 6, 27, 16]
+  screenshot 'index', skip_area: [[17, 6, 27, 16], 'input.-to-skip', '.footer']
```
Comment (donv) [inline README.md:455]: Typo?  Should be 'input.to-skip' ?

**Review by donv — APPROVED** [review]
Looks good to me.

If you would like to use [left, top, width, height] as an argument to `crop`, I am not opposed.

**Comment by pftg** [comment]
> Looks good to me.
> 
> If you would like to use [left, top, width, height] as an argument to `crop`, I am not opposed.

thanks, next time maybe because:

1. after encapsulation things become safer
2. in my projects, will move to use classes first that will simplify migration refactoring

File: README.md (hunk 2 above)
Comment (pftg) [reply README.md:455]: oh sorry this compressed BEM notation, will remove dash

**Pull request merged**